### ISSUE TITLE
OOP Perks [WIP]

### ISSUE THREAD

**Issue by Anonymous-BCFED** [OXOIndustries/Trials-in-Tainted-Space#972]
# tl;dr
The Good:
* I have refactored perks into a set of PerkData classes to make it easier to author perks and mesh them with the game using a simplified event system.
* Perks are managed from a central PerkRegistry class with static members containing each PerkData, as well as an Array of all known perks.
* Perks storage remains the same (StorageClasses in each Creature), but the logic has been centralized.
* **They work.**
* I have an extremely complex set of scripts to 
 * Parse out these perks from the TiTS master branch whenever new perks are made or updated and keep those classes updated.
 * Change all hasPerk() and createPerk() references to PerkRegistry.WHATEVER.* equivalents.
 * Update PerkRegistry with the list of Perks that need to be initialized and recognized.

The Bad:
* The changes are incredibly complex and touch a truckload of code. They won't interfere much with concurrent changes to the code, so this won't block anything anyone else makes, but there's just a lot of shit to change.
* The changes are entangled with some other changes I'd need to separate them from, including buildsystem fixes. This is the biggest problem.
* Not all perks have been ported to the new system yet.
* I do not wish to share most of my build scripts.  Some of the tools, like the code bracket checking tool, may be ported in the future.

The ugly:
* In order to make it easier to merge code from master, the generated code is ugly and will not pass muster in a PR.
 * This ugliness can be removed, but only on a final pass.

**Because of the difficulty involved in porting, I would like to know if anyone would be interested in this changeset before I start messing with it.**

# Summary
A year ago, I found this repo and started messing around with it.  While I have little interest in writing content, I do dabble in adding things here and there, and playing with balance.  About the same time as I made #454, I began refactoring perk logic into PerkData classes. I finally came up with the following guidelines:

1. As much perk logic as possible should be centralized into PerkData classes.
2. Perks should not be referenced by string, as it introduces the probability of typos causing bugs (see [Juggernaut](https://github.com/OXOIndustries/Trials-in-Tainted-Space/blob/a283d640a3411307da478873f72675b95a220074/classes/GameData/Perks.as#L90) vs. [Juggernaught](https://github.com/OXOIndustries/Trials-in-Tainted-Space/blob/54b11cf6c31dc7ab9dea0ceb00b15463d66340e1/classes/GameData/CombatContainer.as#L413))
3. Areas where many perks invoke their logic (such as racialPerkUpdateCheck()) should be avoided and replaced with events, if possible and logical.

After a year of procrastination, I have come up with the following (working) system:

# SimpleEvents
I realize that ActionScript has a built-in events system, but I felt it was like trying to hit a fly with a sledgehammer.  A came up with a simpler event-handling system that just collects callbacks and invokes them when needed.  I then sprinkled these events throughout the system, where advantageous.  Creatures sport several events that replace the ugly forests of if-trees that crop up around frequently-hooked areas of the code, like racialPerkUpdateCheck().
```as
// Creature class fields:
public var OnOrgasm:SimpleEvent = new SimpleEvent(); // Added to ignoredFields, as well.

// In PerkData class:
override public function Attach(c:Creature):void {
    c.OnOrgasm.Attach(this, OnOrgasm); // Attach OnOrgasm to the Creature on Perk application/deserialization.
}
private function OnOrgasm(c:Creature):void {
    trace(perkName+" - OnOrgasm!")
}
// ... in Creature.orgasm():
OnOrgasm.Invoke(this); // Invokes OnOrgasm with a single Creature argument.
```

In this manner, SimpleEvents create an environment that makes it easier to attach shit to in-game events in a modular way.  There are a few weird issues with serialization (namely the `_ignoredFields` array being ignored for no discernible reason), but I have a few hacks in place to make that work, for the time being. 

# PerkData
Instead of being spread all over hell and gone, perk logic is now centralized into PerkData to keep things tidy and maintainable.  Originally, keeping all this code in sync was a pain in the ass, so I only really updated the code for the few perks I was interested in, leaving the older code in-place.  I now have a complex script that uses magic (and special comment tags) to automatically extract and update every perk when Fen updates something.  This obviously won't be in the final edition, but means it'll be easier to stay abreast of changes from upstream if we decide to go forward with this.

Here's what a perk looks like, with annotations:

```as
/* AUTOGENERATED.  **INCLUDE @NO AUTOGEN@ AND DELETE THIS LINE TO DISABLE OVERWRITE.**
 * Updated 10/11/16
 * Invoked:
 * - .\classes\Items\Transformatives\Cerespirin.as:505: target.createPerk("Resin", 0.40, 0, 0, 0, "Your heavy bark skin slows you all the way down, but provides plenty of protection. It also produces resin. Some flighty, energetic, needy species might find it difficult to resist wanting to plaster themselves all over you...");
 * Checked:
 * - .\classes\Creature.as:4006: if(hasPerk("Resin")) scalar = perkv1("Resin");
 * - .\classes\GameData\CombatAttacks.as:452: return (kGAMECLASS.pc.hasPerk("Resin"));
 * - .\classes\Items\Transformatives\Cerespirin.as:122: if(!target.hasPerk("Resin") && target.skinType == GLOBAL.SKIN_TYPE_BARK)
 * - .\includes\game.as:2586: if(pc.hasPerk("Resin"))
 */

package classes.GameData.PerkClasses {
	import classes.Characters.PlayerCharacter;
	import classes.Creature;
	import classes.GLOBAL;
	import classes.GameData.PerkData;
	import classes.StorageClass;
	import classes.kGAMECLASS;

	public class Resin extends PerkData {
		public function Resin() {
			perkName = "Resin";
			perkDescription = "Your heavy bark skin slows you all the way down, but provides plenty of protection. It also produces resin. Some flighty, energetic, needy species might find it difficult to resist wanting to plaster themselves all over you...";
			setStorageValues(0, 0, 0, 0);
			// This stuff is really only used for the save editor (yes, it can fuck with saves now), although it also serves as documentation.
			setStorageName(0, "Max Reflexes multiplier");
			setStorageName(1, "???");
			setStorageName(2, "???");
			setStorageName(3, "???");
		}

		// Called when the perk is loaded into a creature or added.  Events are generally hooked here.
		override public function Attach(c:Creature):void {
			super.Attach(c);
			c.OnPerkRacialUpdateCheck.Attach(this, OnPerkRacialUpdateCheck);
		}

		// Inverse of above.
		override public function Detach(c:Creature):void {
			super.Detach(c);
			c.OnPerkRacialUpdateCheck.Detach(this, OnPerkRacialUpdateCheck);
		}

		override public function Activate(c:Creature):void {
			super.Activate(c);
			// Called when added to a creature.
		}

		override public function Deactivate(c:Creature):void {
			super.Deactivate(c);
			// Called when removed from a creature.
		}

		// Used for cloning the object. May be redundant now, idk
		override public function clone(sc:StorageClass):PerkData {
			var pd:Resin = new Resin();
			return _clone(sc, pd);
		}

		// Automatically generated (definition at .\includes\game.as:2588)
		private function OnPerkRacialUpdateCheck(pc:PlayerCharacter):void {
			trace(perkName+" - OnPerkRacialUpdateCheck!")
			if(pc.skinType != GLOBAL.SKIN_TYPE_BARK)
			{
				kGAMECLASS.eventBuffer += "\n\n" + kGAMECLASS.logTimeStamp("passive") + " The surface of your body tingles and your nose briefly catches a whiff of a familiar amber aroma--which then completely dissipates into the air. Curious, you check your codex and, sure enough, due to the lack of your once bark skin, you’ve lost the ability to create a resin cast to protect yourself. Well, at least you feel a bit more nimble now...";

				kGAMECLASS.eventBuffer += "\n\n(<b>Perk Lost: Resin</b>)";
				pc.removePerk("Resin");
			}
		}

	}
}
```
And yes, all of that (except for a comment here and there) is automatically generated and formatted.  Even the imports are dynamically added and sorted.  I'm weird like that.

## Usage
```as
// Checking for a perk
if(PerkRegistry.PERK.hasPerk(creature)) ...
// Adding a perk
PerkRegistry.PERK.applyTo(creature);
// Removing a perk
PerkRegistry.PERK.removeFrom(creature);
```
Simple, and the enum-like structure will make the compiler flip its shit if you fat-finger a perk's name.  Some perks also have methods for manipulating StorageClass values. (`PerkRegistry.NUKI_NUTS.SetBonus(creature, 0)`)

# WHY?
The entire purpose of going through all this mayhem is to make it easier on you guys in the future.  You won't have to hunt down every last mention of "'Nuki Nuts" to change a few values, or run around in multiple files adding hooks to common functions.  For the most part, you just throw in a new class, add the event hooks you need to it, and add it to the registry.  You're done. You also get the benefits of OOP, and the events system can be used for a bunch of other things in the future.
